### PR TITLE
[Image] Add reader for numpy .npy images

### PR DESCRIPTION
**Pull request by vuzelac-cadence**
**Summary:**
Allow loading numpy .npy format encoded images. Currently PNG loader is extended to recognize the header and call PNG or numpy loader.

**Documentation:**
Numpy format Limitations
Only little-endian is supported.
FORTRAN data ordering is not supported.
Format v2.0 and v3.0 are not supported.

**Input files layout information.**
Tensors can be assigned an image layout using` -input-layout=LAYOUT` command line argument. Layout can be:

- An image layout - NHWC or NCHW. In this case, tensors must be 3D or 4D only. 3D ones are expanded with the batch. If the layout is not matching the one specified using existing `-image-layout `command line argument, the input tensor is transposed accordingly. Tensors from multiple files are concatenated along the batch dimensions. Thus, the batch dimension can differ from file to file but other dimensions must match.

- NonImage  layout (`-image-layout=NonImage`). Can be uses for 1D-4D tensors. In this case,` -input-layout` is ignored, input files are not transposed, and batching is not possible. Also, these tensors accept a single mean/stddev only.

**Data preprocessing**
Tensors are treated the same as images. That is, existing switches -stddev, -mean and -image-mode apply. When assigning sttdev or mean, have to use as many channels as there is in the tensors (except for `-image-layout=NonImage` where a single stddev/mean is expected).

**Loading multiple files**
Tensors from multiple files are concatenated along the batch dimensions (not applicable with `-image-layout=NonImage`). Thus, the batch dimension can differ from file to file but other dimensions must match.

**Glow command line**
Existing `-input-image-file-list` or positional file arguments are used. NUMPY format is automatically determined. The only addition is `-input-layout` command line switch described above.

Test Plan:
ImageTest extended with relevant tests

### CODE REVIEW

**Comment by vuzelac-cadence** [comment]
@jfix71 this is the second attempt to implement numpy reader. Previous one PR #4554  was unfortunately closed. I promise to carry out the PR this time. Your comments from the old PR are incorporated, however, I'd like to discuss one remaining issue:

> So we always use data with float -- do you have private uses for other datatypes? Wondering because buffertoTypedVector() has support for other types but doesn't appear we're using/testing that anywhere.

I didn't quite understand this - we don't use float but U8 - for PNG for example it's U8. It's just Glow that converts that to float. Realistically, we cannot even use the float data; after we generate the optimized code for our NN engine we will work with U8 and S8 data for example. So, the use of U8 is not private but as general as using the PNG,

Now, if you look at the new code, we do support U8 and S8 and just convert them to float. This is all that we care about and I removed some complicated logic I had there. We don't need more than this. I would assume  that whoever needs more can freely add. 
We do have some internal Glow changes that insert BatchNorm instead of using the normalization code in Image.cpp or NumpyReader.cpp. The implementation currently limits us to U8 only. We do support S8 input data by adding +128 to the values, and then we expect users to actually fix their stddev/mean values.
Here, `convertU8S8ToFloat `I still contains our own internal code, just to show you that we do need a separate handler for INT8 data type, but the final code is to have +128 removed. I'll fix that after I get some feedback from you on other changes you'd like to see here.

**Comment by vuzelac-cadence** [comment]
@jfix71  ../include/glow/Base/Image.h:20:2: error: "Using Glow's PNG library requires installing libpng"
Seems like some internal problem

**Comment by vuzelac-cadence** [comment]
Hope to get this reviewed soon.

**Comment by jfix71** [comment]
Will review tomorrow. Sorry about all the delays.

File: lib/Base/NumpyReader.cpp [hunk 1]
```diff
@@ -0,0 +1,302 @@
...
+      dataNpy.shape.insert(dataNpy.shape.begin(), 1);
+    }
+    CHECK_EQ(dataNpy.shape.size(), 4)
+        << "NPY loader: Supporting only 3 or 4 dimensions.";
+    normalizeData(imageLayout, mean, stddev, imageNormMode, data,
+                  dataNpy.shape);
+    // Load tensor from the vector obtained from the npy loader.
+    tensors[n].reset(glow::ElemKind::FloatTy, dataNpy.shape);
+    tensors[n].getHandle<>() = data;
+    auto dims0 = tensors[0].dims();
```
Comment (jfix71) [inline lib/Base/NumpyReader.cpp:257]: nit: move this outside the loop?

File: lib/Base/NumpyReader.cpp [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+    if (n > 1) {
+      CHECK_EQ(dims0.size(), dims.size())
+          << "NPY loader: Number of dimensions must match.";
+      for (dim_t i = 1; i < dims.size(); i++) {
+        CHECK_EQ(dims0[i], dims[i])
+            << "NPY loader: Non-batch dimensions must match.";
+      }
+    }
+    // Accumulate batch dimension after each dump is loaded.
+    batchSize += tensors[n].dims()[0];
```
Comment (jfix71) [inline lib/Base/NumpyReader.cpp:268]:
```suggestion
    batchSize += dims[0];
```

File: lib/Base/NumpyReader.cpp [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    CHECK_EQ(dataNpy.shape.size(), 4)
+        << "NPY loader: Supporting only 3 or 4 dimensions.";
+    normalizeData(imageLayout, mean, stddev, imageNormMode, data,
+                  dataNpy.shape);
+    // Load tensor from the vector obtained from the npy loader.
+    tensors[n].reset(glow::ElemKind::FloatTy, dataNpy.shape);
+    tensors[n].getHandle<>() = data;
+    auto dims0 = tensors[0].dims();
+    auto dims = tensors[n].dims();
+    if (n > 1) {
```
Comment (jfix71) [inline lib/Base/NumpyReader.cpp:259]: Shouldn't this be `n > 0` ?

File: lib/Base/NumpyReader.cpp [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    batch += tensors[n].dims()[0];
+  }
+
+  // Convert to the requested layout.
+  if (inputLayout != imageLayout && inputLayout != ImageLayout::Unspecified) {
+    glow::Tensor transposed;
+    if (imageLayout == ImageLayout::NCHW) {
+      inputData.transpose(&transposed, {0u, 3u, 1u, 2u});
+    } else {
+      inputData.transpose(&transposed, {0u, 2u, 3u, 1u});
```
Comment (jfix71) [inline lib/Base/NumpyReader.cpp:288]: Can we move `0u, 3u, 1u, 2u` and other into a `#define NHWC2NCHW`, and reuse elsewhere in this file? As is done in Graph.h (I'm assuming you didn't want to include that here because of a dependency issue?)

File: lib/Base/NumpyReader.cpp [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+struct NpyData {
+  std::vector<char> data;
+  std::vector<dim_t> shape;
+  NpyType type;
+  size_t nvals;
+  size_t elemSize;
+};
+
+void convertU8S8ToFloat(NpyData &dataNpy, std::vector<float> &data) {
```
Comment (jfix71) [inline lib/Base/NumpyReader.cpp:44]: nit: Add doxygen comment.

**Review by jfix71 — COMMENTED** [review]
@vuzelac-cadence Ok! Finally reviewed, apologies.

I appreciated the great explanation in the description for the PR -- it would be great to move some of that into the descriptions for the command line options, or at least as comments in the code.

File: lib/Base/NumpyReader.cpp (hunk 4 above)
Comment (vuzelac-cadence) [reply lib/Base/NumpyReader.cpp:288]: I didn't include Graph.h as it seem not to belong here - here I am working with Tensors only and some image processing stuff, thus no need to include anything else.  The transpose formats maybe should be in Tensor.h ? Otherwise, I can replicate here

File: lib/Base/NumpyReader.cpp (hunk 1 above)
Comment (vuzelac-cadence) [reply lib/Base/NumpyReader.cpp:257]: Didn't understand why - performance wise should be the same as aliasing rules will decide if the compiler will move it out of loop itself. Coding style ? Note that tensor is not yet created there.

File: lib/Base/NumpyReader.cpp (hunk 3 above)
Comment (vuzelac-cadence) [reply lib/Base/NumpyReader.cpp:259]: Thanks!

**Comment by vuzelac-cadence** [comment]
> it would be great to move some of that into the descriptions for the command line options, or at least as comments in the code.

I didn't get this - some .md document or the descriptiion of some cl::opt ? 
I'd assume the PR description better go to some MD but not sure which one ?

**Comment by vuzelac-cadence** [comment]
@jfix71 , this keeps randomly failing on ThreadPoolExecutorTest

**Comment by vuzelac-cadence** [comment]
Hi @jfix71 , hope you can look at this.

**Comment by vuzelac-cadence** [comment]
ping

File: lib/Base/NumpyReader.cpp (hunk 1 above)
Comment (jfix71) [reply lib/Base/NumpyReader.cpp:262]: Ah nevermind, misread it.

**Review by jfix71 — COMMENTED** [review]
Hi @vuzelac-cadence -- think this is in good shape, though I think some of my comments in `NumpyReader.cpp` were never addressed.

>I didn't get this - some .md document or the descriptiion of some cl::opt ?
>I'd assume the PR description better go to some MD but not sure which one ?

cl::opt, yeah. For md, could be at the end of `Testing.md`, something about some options available for loading images for ImageClassifier?

**Comment by vuzelac-cadence** [comment]
@jfix71 added details to Testing.md.
I don't understand why the test fails. Pytorch failed on the previous commit, previously was something else, now it's DEBUG:
ThreadPoolExecutorTest ................***Exception: SegFault  6.28 sec

**Comment by vuzelac-cadence** [comment]
I've asked before, but want to ask again: Can you add triggers for easy relaunch.
What we use although in Gitlab, thus easier,  is a comment with a keyword that triggers rebuild. 
Since you are using CI Workflow this should be even easier - there is  Rerun button - need to give permissions to the author for example (not sure if it's that easy...).

**Review by jfix71 — APPROVED** [review]
LGTM.

RE: triggering retesting -- yeah I think the main issue here would be the permissions. I know we have a `pytorch-bot` in the mainline PyTorch repo where you can comment `@pytorchbot retest this please` and it will relaunch tests for you. Not sure I can try to see if someone has the cycles on our end to enable this.

**Comment by vuzelac-cadence** [comment]
Ok, I tried once again, now 32B_DIM_T failed but it's always ThreadPool test.
